### PR TITLE
Add CATEGORICAL_ERROR cost for one-hot multi-class targets (#2736)

### LINKED ISSUE #2736: evolveDir error/score can reach trivial one-hot MSE floor while classification accuracy stays at chance
## Summary

For multi-class supervised problems encoded as **one-hot targets** in a binary `.bin` stream, `evolveDir` can report a low training `error` (and high `score`) that **does not correspond to useful argmax classification**. The library stop condition (`targetError`) can therefore declare success while the evolved creature is still at chance accuracy.

This is a **NEAT-AI metric / fitness semantics** issue, not an example-repo orchestration issue.

## Repro (via NEAT-AI-Examples showcase)

Minimal seed (documented NEAT-AI initialisation):

```ts
new Creature(784, 10)
// → 794 neurons (784 inputs + 10 outputs), ~7k synapses
```

```bash
./mnist_classification/run.sh --timeout=15
# Resume across many chunks without --fresh
```

## Observed (NEAT-AI 5.0.29, 60k-record `.bin`, 10-class one-hot)

| Metric | Typical plateau |
|--------|-----------------|
| `evolveDirError` | **~0.10** (hits `targetError` thresholds) |
| `evolveDirScore` | **~0.90** |
| Held-out **argmax test accuracy** | **~10%** (chance for 10 classes) |
| Neuron count | Stuck at **794** (no hidden structure added) |

Extended campaigns (200+ fifteen-minute chunks, thousands of cumulative generations) show **monotonically decreasing recorded training error** but test accuracy never exceeds ~14% on minimal seed.

## Why this matters for library users

Consumers of `evolveDir` naturally assume `error` measures task quality. For one-hot multi-class targets the **trivial MSE floor is ~0.1** (one output near 1, nine near 0 on average), which can be reached by a constant or near-constant classifier — **without** learning digit structure.

The example measures argmax accuracy separately and finds the disconnect; the library itself does not surface classification quality.

## Expected (library)

At least one of:

1. **Document** that `error`/`score` on one-hot `.bin` streams is MSE-based and can decouple from argmax accuracy for multi-class problems.
2. **Offer** an optional classification-aware error / early-stop metric for multi-class one-hot layouts.
3. **Ensure** champion selection / `targetError` stop reflects the user's actual task when outputs are one-hot class labels.

## Contrast (same example, different seed)

Using NEAT-AI layered initialisation (`new Creature(784, 10, { layers: [{ count: 64, squash: "ReLU" }] })`) in the same pipeline:

- `evolveDirError` ≈ **1.33** after one 15-minute chunk (not stuck at 0.1)
- Test accuracy ≈ **14.9%**
- Topology grows (858 → 1049 neurons)

So the library **can** learn when given hidden structure; the minimal-seed metric plateau appears to be a fitness-signal problem, not missing backprop or Discovery per se.

## Related

Originally observed in NEAT-AI-Examples [#446](https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/446) (closed) — filing here because the fix belongs in NEAT-AI.

### PR DESCRIPTION
**Pull request by stservice**
# Add `CATEGORICAL_ERROR` cost for one-hot multi-class targets

## Summary

For multi-class problems encoded as **one-hot targets**, `evolveDir` could
report a low `error` / high `score` (and even satisfy `targetError`) while real
argmax classification accuracy stayed at chance. The cause is metric semantics,
not training: distance-based costs such as `"MSE"` measure squared output
distance, not argmax accuracy. A constant classifier that emits one output near
`1` and the rest near `0` reaches a trivial MSE floor (~`0.1` for ten balanced
classes) without learning any class structure. Because
`score = 1 - error - penalties`, that decoupling lets the reported score
overstate task quality.

This PR adds a new built-in cost, **`"CATEGORICAL_ERROR"`**, that reports the
**misclassification rate** (`error = 1 - accuracy`) by comparing
`argmax(target)` with `argmax(output)` per record. Selecting it makes the
scorer, champion selection and the `targetError` early-stop reflect the user's
actual classification task. The cost is **non-differentiable** and intended for
scoring/early-stop — NEAT-AI derives backpropagation gradients independently of
`costName`, so training proceeds normally. The behaviour and the MSE-decoupling
caveat are documented in `docs/API_REFERENCE.md`.

This satisfies the issue's expected outcomes: it both **documents** the
MSE/argmax decoupling and **offers an optional classification-aware metric** so
champion selection / `targetError` can reflect the real task.

Closes #2736

## Evidence

Backend/library change with no web interface to screenshot. Verified with unit
tests that call the real cost function and the `Costs` registry.

```mermaid
flowchart LR
    T[One-hot target] --> A[argmax target]
    O[Creature output] --> B[argmax output]
    A --> C{equal?}
    B --> C
    C -- yes --> Z[0 correct]
    C -- no --> N[1 incorrect]
    Z --> M[mean over dataset = 1 - accuracy]
    N --> M
    M --> S["score = 1 - error - penalties<br/>now reflects argmax accuracy"]
```

Key behaviours verified by tests:

- A constant ten-class classifier yields `error = 0.9` (chance accuracy 10%),
  versus the misleading ~`0.1` MSE floor.
- A perfect classifier yields `error = 0`.
- Correct/incorrect single-record predictions return `0` / `1`.
- Ties resolve to the first index; empty arrays return `0`; a single output is
  always counted correct (documented limitation).
- `Costs.find("CATEGORICAL_ERROR")` resolves and the name appears in
  `Costs.getAvailableCosts()` and `BUILT_IN_COST_NAMES`.

## Test Plan

- Added `test/costs/CategoricalError.ts` (9 tests) covering happy path, error
  path, registry integration, the one-hot MSE-floor scenario, and edge cases
  (ties, empty, single output).
- Updated `test/costs/CostName.ts` so `BUILT_IN_COST_NAMES` stays in sync with
  the new built-in cost.
- All `test/costs/*.ts` tests pass (63 passed, 0 failed).
- `./quality.sh --lint-only` passes; changed files type-check cleanly with
  `deno check`.

> Note: `./quality.sh --check-only` reports 3 pre-existing `TS2322`
> `Timeout`/`number` errors in `DataRecorderAnalysis.ts`,
> `DataRecorderRecording.ts` and `ImproveSquash.ts`. These exist on the base
> branch (confirmed via `git stash`) and are unrelated to this change.

---

🤖 Processed by: stservice

### CODE REVIEW

**Comment by stservice** [comment]
I reviewed the CI check failure (**quality**) but could not produce a code fix. The classifier indicates this requires a code change, so a human reviewer is required — I have added the `needs-human` label.

**Classifier reason:** code-fix pattern matched: ts2322
**Signals:**
- `text:ts2322`
- `text:expected:`

**Review by nleck — APPROVED** [review]
